### PR TITLE
WIP:improve zoom behavior Presentation

### LINKED ISSUE #671: Presentation Mode not really working
- Zoom working strange - should be always 100% I think
- Toolbar should be selectable in the settings
- Create a Presentation toolbar, and select as default

### PR DESCRIPTION
**Pull request by peetCreative**
This rather bit too big pullrequest is supposed to fix various problems with zoom and view modes,
as well refactors zoom related parts:
Fix #671 #772
I would be happy, if someone has a look on it.

Präsentationmode:
- Zoom and Scroll is disabled
- Zoom that whole current page is viewed
- disable Paired, Rows or Columns settings
- Navigation with Keys: Up/next page,  down/previous page
- Zoom action Buttons are disabled instead of hidden 
Zoom Fit:
-  Zoom is adjusted by listining size-allocate signal of widget
Problems/Todo (to be created issues if merge):
- Presentation icon has to be designed. presentation-mode.svg is nothing saying. I'm not a graphic designer
- Sometimes after disabling Presentation-mode scroll is still disabled
- Layout::scrollTo does not work correctly.
-- Fitmode enabled and several columns, rows or paired: when going from one to the next page horizontal scroll is not working
-- Presentationmode and Document with horizontal differently sized pages gives same problems

### CODE REVIEW

**Comment by LittleHuba** [comment]
Very nice to get this view mode actually running! I'll take a look sometime later.

**Comment by JJones780** [comment]
### re: Layout::scrollTo does not work correctly.
XournalView::scrollTo  calls layout::ensureRectVisible.

In `void XournalView::scrollTo(size_t pageNo, double yDocument)`
 the yDocument argument is outdated and should be 0.

I'm slowly working through some cleanup leftover from the multilayout change and this is one of them.
If you want to leave the scrollTo issue and merge this code as is, I will put the fix above in with my changes.
I will also look at the other scrollTo issues you mentioned as soon as you're merged. I think some may already be fixed in my working branch.

**Comment by JJones780** [comment]
re: zoom

Also, I've rewritten Layout::UpdateCurrentPage and find that zooming behaves better if this function does not select a new page in the visible area. My update will reflect this I expect. 
I don't think this affects your existing changes.

**Comment by JJones780** [comment]
Does presentation mode require changing the layout of the document?
If so we should really create a 
void LayoutMapper::configureForPresentation(int numPages) which calls the generic  LayoutMapper::configure with some fixed settings  which is called when presentation mode is activated. 
Then when presentation mode is deactivated the user gets their original layout back.
Also, disabling the layout menu items during presentation mode would avoid confusion.
But:
I'd suggest considering NOT changing the layout for presentation.

**Comment by peetCreative** [comment]
I will have a look at your comments tonight

File: src/control/Control.h [hunk 1]
```diff
@@ -126,10 +126,8 @@ class Control :
 	void saveSettings();
 
 	void updateWindowTitle();
-	void calcZoomFitSize();
-	void setZoomFitButton(bool enabled);
 	void setViewPairedPages(bool continous);
-	void setViewPresentationMode(bool continous);
+	void setViewPresentationMode(bool);
```
Comment (andreasb242) [inline src/control/Control.h]: Keep the var name also in the prototype, for documenation

File: src/control/zoom/ZoomControl.cpp [hunk 2]
```diff
@@ -27,12 +27,18 @@ void ZoomControl::zoomOneStep(bool zoomIn, double x, double y)
 {
 	XOJ_CHECK_TYPE(ZoomControl);
 
+	if(this->zoomPresentationMode)
```
Comment (andreasb242) [inline src/control/zoom/ZoomControl.cpp]: This probably solves the ticket: Presentation mode not working?!

File: src/gui/MainWindow.cpp [hunk 3]
```diff
@@ -532,32 +531,6 @@ bool MainWindow::onKeyPressCallback(GtkWidget* widget, GdkEventKey* event, MainW
 		//editing text - give that control
 		return false;
 	}
-	else if (event->keyval == GDK_KEY_Down)
```
Comment (andreasb242) [inline src/gui/MainWindow.cpp]: Why is this removed?

File: src/gui/MainWindow.cpp [hunk 4]
```diff
@@ -576,17 +549,6 @@ bool MainWindow::deleteEventCallback(GtkWidget* widget, GdkEvent* event, Control
 	return true;
 }
 
-bool MainWindow::configureEventCallback(GtkWidget* widget, GdkEventConfigure* event, Control* control)
```
Comment (andreasb242) [inline src/gui/MainWindow.cpp]: Why is this removed, was this dead code from GTK2? ("configure-event")

File: ui/icons/hicolor/scalable/actions/presentation-mode.svg [hunk 5]
```diff
@@ -0,0 +1,57 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (andreasb242) [inline ui/icons/hicolor/scalable/actions/presentation-mode.svg]: As there is no about comment: You checked the license of the icon or created it by yourself?

Comment (peetCreative) [reply ui/icons/hicolor/scalable/actions/presentation-mode.svg]: I created a new logo, but it's really nothing saying and I would request someone to create a new one.

File: src/gui/MainWindow.cpp (hunk 3 above)
Comment (peetCreative) [reply src/gui/MainWindow.cpp]: The same is done in XournalView.cpp:235 but this is called first.

File: src/control/zoom/ZoomControl.cpp (hunk 2 above)
Comment (peetCreative) [reply src/control/zoom/ZoomControl.cpp]: Well, it's just disables Zoom during presentationMode.
Maybe we should do this rather in zoomSequenceChanges?

File: ui/icons/hicolor/scalable/actions/presentation-mode.svg (hunk 5 above)
Comment (andreasb242) [reply ui/icons/hicolor/scalable/actions/presentation-mode.svg]: The question was only about the license...

Comment (peetCreative) [reply ui/icons/hicolor/scalable/actions/presentation-mode.svg]: I would just say CC, I will push sth soon.

File: src/gui/MainWindow.cpp (hunk 4 above)
Comment (peetCreative) [reply src/gui/MainWindow.cpp]: I added this some commits before but it didn't catched when the application was resized using tiling. I moved to the signal size-allocate from the widget and the ZoomControl is listening to it. Now everything is working.

File: src/control/Control.cpp [hunk 6]
```diff
@@ -1523,50 +1529,43 @@ void Control::changePageBackgroundColor()
...
-	settings->setPresentationMode(presentationMode);
-	fireActionSelected(GROUP_PRESENTATION_MODE, presentationMode ? ACTION_VIEW_PRESENTATION_MODE : ACTION_NOT_SELECTED);
+	if(enabled)
+	{
+		// Disable PresentationMode if ZoomValue could not calculated correctly
+		zoom->updateZoomPresentationValue();
+		//TODO: write out to settings previous layout
+		setViewPairedPages(false);
+		setViewRows(0);
+		setViewColumns(0);
```
Comment (LittleHuba) [inline src/control/Control.cpp]: Shouldn't that be `setViewColumns(1)` so we have one column for pages?

File: src/control/zoom/ZoomControl.cpp (hunk 2 above)
Comment (LittleHuba) [reply src/control/zoom/ZoomControl.cpp]: I would do it at the earliest convenient position in the code, to decrease the overhead. But essentially it is negligible.

File: src/gui/MainWindow.cpp (hunk 3 above)
Comment (LittleHuba) [reply src/gui/MainWindow.cpp]: XournalView is the better position to do this. As that is actually the view that should react to the event.

File: src/gui/XournalView.cpp [hunk 7]
```diff
@@ -20,6 +20,9 @@
 
 #include <gdk/gdk.h>
 
+#include <tuple>
+#include <math.h>
```
Comment (LittleHuba) [inline src/gui/XournalView.cpp]: You should use cmath instead or we revive a bug previously fixed.

File: src/control/Control.cpp (hunk 6 above)
Comment (JJones780) [reply src/control/Control.cpp]: LittleHuba is correct but now that you've merged in #1037 you should not change the layout settings here.
You should remove the calls to  setViewRows, setViewColumns and setViewPairedPages.
PR 1037 did not configure the Paired Pages so you'll have to remedy that by changing   LayoutMapper::configureForPresentation() so that line
`	bool isPairedPages = settings->isShowPairedPages();`
instead reads:
`	bool isPairedPages = false;`

Comment (JJones780) [reply src/control/Control.cpp]: When the user leaves presentation mode the layout will revert back to using the settings.
Because Presentation mode ignores the users layout settings we should also grey out the layout settings ( View->Layout and view->Cols/Rows ) in the menu when presentation mode is active. 
_Can do in another PR._

File: src/gui/XournalView.cpp [hunk 8]
```diff
@@ -615,14 +616,25 @@ void XournalView::zoomChanged()
...
 	layout->layoutPages();
 
-	zoom->scrollToZoomPosition(view);
+	if(zoom->isZoomPresentationMode() || zoom->isZoomFitMode())
+	{
+		scrollTo(currentPage);
+	}
+	else
+	{
+		std::tuple<double, double> pos = zoom->getScrollPositionAfterZoom(view);
```
Comment (JJones780) [inline src/gui/XournalView.cpp]: You can use std:tie here to be a bit more readable:
		int  scrollX , scrollY;
		std::tie( scrollX , scrollY ) = zoom->getScrollPositionAfterZoom(view);
		if( scrollX != -1 &&  scrollY != -1)
		{
			layout->scrollAbs(  scrollX , scrollY );
		}

File: src/gui/XournalView.cpp [hunk 9]
```diff
@@ -615,14 +616,25 @@ void XournalView::zoomChanged()
...
+		if(std::get<0>(pos) != -1 && std::get<1>(pos) != -1)
+		{
+			layout->scrollAbs(std::get<0>(pos), std::get<1>(pos));
+		}
+	}
```
Comment (JJones780) [inline src/gui/XournalView.cpp]: You can use std:tie here to be a bit more readable:
```
int scrollX , scrollY;
std::tie( scrollX , scrollY ) = zoom->getScrollPositionAfterZoom(view);
if( scrollX != -1 && scrollY != -1)
{
layout->scrollAbs( scrollX , scrollY );
}
```

**Comment by JJones780** [comment]
I can't seem to be able to use the pgup or pgdown keys to go more than one page away?
Also, it doesn't zoom in on the current page when I go into Presentation mode?

File: src/gui/XournalView.cpp [hunk 10]
```diff
@@ -451,9 +455,6 @@ void XournalView::scrollTo(size_t pageNo, double yDocument)
 	int height = v->getDisplayHeight();
 
 	layout->ensureRectIsVisible(x, y, width, height);
-
-	// Select the page
-	control->firePageSelected(pageNo);
```
Comment (JJones780) [inline src/gui/XournalView.cpp]: Restore this... we need to firePageSelected here! 
Page navigation won't work without it.

Comment (JJones780) [inline src/gui/XournalView.cpp]: Restore this... we need to firePageSelected here!
Page navigation won't work without it.

File: src/control/Control.cpp (hunk 6 above)
Comment (LittleHuba) [reply src/control/Control.cpp]: @JJones780 we should do this in another PR as this one is already heavy enough. Additionally it is coming up with Merge Conflicts where we should not introduce another feature in fear of even more conflicts.

Comment (peetCreative) [reply src/control/Control.cpp]: I disabled the menuitems now, this should be enough.

File: src/gui/XournalView.cpp (hunk 10 above)
Comment (peetCreative) [reply src/gui/XournalView.cpp]: reverted it.

File: src/control/Control.cpp [hunk 11]
```diff
@@ -1523,50 +1533,67 @@ void Control::changePageBackgroundColor()
...
-void Control::setViewPairedPages(bool pairedPages)
-{
-	XOJ_CHECK_TYPE(Control);
+		setViewLayoutVert(settings->getViewLayoutVert());
+		setViewLayoutR2L(settings->getViewLayoutR2L());
+		setViewLayoutB2T(settings->getViewLayoutB2T());
 
-	settings->setShowPairedPages(pairedPages);
-	fireActionSelected(GROUP_PAIRED_PAGES, pairedPages ? ACTION_VIEW_PAIRED_PAGES : ACTION_NOT_SELECTED);
+		setViewPairedPages(settings->isShowPairedPages());
```
[the remark this answers is not in the document]
Comment (peetCreative) [reply src/control/Control.cpp]: It is meant to restore the the Layout, Cols/Rows after coming back from Presentationmode

File: src/control/Control.cpp [hunk 12]
```diff
@@ -1523,50 +1533,67 @@ void Control::changePageBackgroundColor()
...
-}
+	fireActionSelected(GROUP_PRESENTATION_MODE, enabled ? ACTION_VIEW_PRESENTATION_MODE : ACTION_NOT_SELECTED);
+	//Disable Zoom
+	fireEnableAction(ACTION_ZOOM_IN, !enabled);
+	fireEnableAction(ACTION_ZOOM_OUT, !enabled);
+	fireEnableAction(ACTION_ZOOM_FIT, !enabled);
+	fireEnableAction(ACTION_ZOOM_100, !enabled);
+	fireEnableAction(ACTION_FOOTER_ZOOM_SLIDER, !enabled);
+	//Disable Viewmodes without unsetting them
+	fireEnableAction(ACTION_VIEW_PAIRED_PAGES, !enabled);
```
[the remark this answers is not in the document]
Comment (peetCreative) [reply src/control/Control.cpp]: Well this is just supposed to disable the button and menu for zoom.
I have to check the paired pages behavior, while presentation.

**Comment by peetCreative** [comment]
>     * The zoom behaviour is also not right.. it scrolls and zooms simultaneously ( when using control with the mouse wheel ).

I can't reproduce this.

**Comment by peetCreative** [comment]
> When setting presentation mode it doesn't scroll to the proper page. There seem to be other issues with page numbers so this is likely not associated with your changes.

I noticed that the current focused page is not updated anymore. So if one is just scrolling further and not drawing or sth. The current page is not updated. And if then Presentation is set it goes back to the first or previous used page, which is not at all visible.

**Comment by peetCreative** [comment]
There are also lots of issues with paired pages in presentation and fit zoom, when the pages are not equally width-sized.

**Comment by peetCreative** [comment]
I also noticed that size-allocate of the xournalview GTK widget is not thrown consistently and is sometimes working sometimes not and changing working on setting Presentationmode. Very strange.
